### PR TITLE
HDDS-10486. Recon datanode UI to incorporate explicit removal of DEAD

### PR DESCRIPTION
**Pull request by smitajoshi12**
## What changes were proposed in this pull request?
Recon datanodes listing page should have a mechanism using a button or link which will be enabled once user selects the  State is **DEAD** nodes for explicity removal from Recon memory and ask recon to stop tracking. 

**Node State= DEAD    Operation State=[Any State] ==> Allow for Removal**

Following backend API will be provided by Recon to facilitate this.

1) When totalcount is greater than 0 then we are displaying Remove button.

Please describe your PR in detail:
PUT
api/v1/datanodes/remove
JSON Body input takes UUID list of datanodes to be removed from cluster:
Input:
[
"50ca4c95-2ef3-4430-b944-97d2442c3daf"
]
Output:

Examples of well-written pull requests:

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-10486

## Depending on Backend PR
Need to get merge after  below backend  PR gets merged 
https://github.com/apache/ozone/pull/6360

https://issues.apache.org/jira/browse/HDDS-11032)

## How was this patch tested?
Manually

Only Dead node removal

image

image
-----------------------------------------------------------------------------------------------------------
Dead Node Removal
Screenshot 2024-05-02 at 5 37 43 PM

Screenshot 2024-05-02 at 5 37 43 PM
----------------------------------------------------------------------------------------------------------
In Case of failure and abort signal
Screenshot 2024-05-30 at 5 26 06 PM

### CODE REVIEW

**Comment by ArafatKhan2198** [comment]
When the `ozone-datanode-2.ozone_default` becomes DEAD, I am unable to select and remove it. Is there a waiting period for it, or am I missing something here?

image

**Comment by ArafatKhan2198** [comment]
When no nodes are selected, the Remove button remains deactivated, which is the correct behavior. However, upon clicking it, I still receive a popup message indicating the deletion of zero nodes. Please refrain from displaying the popup message and making the API call when no nodes are selected.

image

**Comment by ArafatKhan2198** [comment]
Additionally, please reduce the size of the popup message and align it properly to avoid covering any fields. I suggest reducing its size slightly and aligning it appropriately within the available space.

**Comment by ArafatKhan2198** [comment]
Also, please ensure that the "?" icon in the alert message is properly aligned, similar to the alert message below.

image

**Comment by ArafatKhan2198** [comment]
And do test it out not just with `DB.json` but also an actual cluster, for getting a `DEAD` node just shut down one of the Datanodes.

**Review by ArafatKhan2198 — CHANGES_REQUESTED** [review]
Thanks for working on this @smitajoshi12 please find a few comments I have left.

**Comment by smitajoshi12** [comment]
> When the `ozone-datanode-2.ozone_default` becomes DEAD, I am unable to select and remove it. Is there a waiting period for it, or am I missing something here?
> 
> image

@devmadhuu 
Can we remove Dead Node status UI is designed for Decomiisioned and In Maintenance state removeal only?

**Review by devmadhuu — COMMENTED** [review]
Thanks @smitajoshi12 for working on the patch. Kindly check below points:

1. "Remove" button looks ugly. Pls make its layout and font inline with UX of Recon. Try using some image , red color 'X' symbol image , may be small icon and add tooltip to describe the purpose of that icon.
2. As discussed, add an alert only when some nodes are selected and then, when user clicks on remove icon. Keep UX and look and feel of alert proper.
3. We need to also allow removal of "DEAD" state datanodes. Currently it is not allowing.

**Comment by devmadhuu** [comment]
> > When the `ozone-datanode-2.ozone_default` becomes DEAD, I am unable to select and remove it. Is there a waiting period for it, or am I missing something here?
> > image
> 
> @devmadhuu Can we remove Dead Node status UI is designed for Decomiisioned and In Maintenance state removeal only?

Yes, we need to allow removal of "DEAD" state datanodes. Currently it is not allowing.

**Comment by smitajoshi12** [comment]
> When no nodes are selected, the Remove button remains deactivated, which is the correct behavior. However, upon clicking it, I still receive a popup message indicating the deletion of zero nodes. Please refrain from displaying the popup message and making the API call when no nodes are selected.
> 
> image

@ArafatKhan2198 
Addressed issue in latest commit and latest screen shots.

**Comment by smitajoshi12** [comment]
> Also, please ensure that the "?" icon in the alert message is properly aligned, similar to the alert message below.
> 
> image

@ArafatKhan2198 
It is axios error will raise seprate JIRA for this task.

**Comment by smitajoshi12** [comment]
> > When the `ozone-datanode-2.ozone_default` becomes DEAD, I am unable to select and remove it. Is there a waiting period for it, or am I missing something here?
> > image
> 
> @devmadhuu Can we remove Dead Node status UI is designed for Decomiisioned and In Maintenance state removeal only?

@ArafatKhan2198 
Completed  changes

**Comment by devmadhuu** [comment]
@smitajoshi12 thanks for continuing working on this patch, Image icon for remove click seems not correct, as it is showing as "i" informative icon.

**Comment by dombizita** [comment]
@smitajoshi12 I marked this PR as draft, while we are waiting for the backend changes to be merged in #6360

**Comment by smitajoshi12** [comment]
> @smitajoshi12 thanks for continuing working on this patch, Image icon for remove click seems not correct, as it is showing as "i" informative icon.

@devmadhuu 
Updated with Current Cluster data screen shots.

**Comment by ArafatKhan2198** [comment]
Thanks for updating the patch @smitajoshi12 
Can you please fix the message in the tool tip to "Remove and stop tracking the DECOMMISSIONED, IN_MAINTENANCE, and DEAD nodes." 

image

**Comment by ArafatKhan2198** [comment]
Just one comment to address that I have posted, and we should be good to merge this PR.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx:429]: Could we update this to use the axios helpers? Since this is a PUT operation it would be better we ensure the request is cancelled if we are changing context.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx:427]: Just a nit, can we add a space before the `=`.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx:432]: Since
```
this.setState({
        loading: false,
        selectedRowKeys: []
});
```
is being called in both the catch and try statement as the final state change, can we instead shift it to a `finally` block?

**Comment by smitajoshi12** [comment]
> Thanks for updating the patch @smitajoshi12 Can you please fix the message in the tool tip to "Remove and stop tracking the DECOMMISSIONED, IN_MAINTENANCE, and DEAD nodes."
> 
> image

@ArafatKhan2198 
What should be exact message?

**Comment by ArafatKhan2198** [comment]
> Thanks for updating the patch @smitajoshi12 Can you please fix the message in the tool tip to "Remove and stop tracking the DECOMMISSIONED, IN_MAINTENANCE, and DEAD nodes."
> 
> image

The message should be "Remove and stop tracking the DECOMMISSIONED, IN_MAINTENANCE, and DEAD nodes." without the extra exclamation marks at the end.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: @devabhishekpal  completed changes in latest commit.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: @devabhishekpal 
Completed changes in latest commit and tested locally and attached screenshot for abort signal also

**Comment by smitajoshi12** [comment]
> Thanks for updating the patch @smitajoshi12 Can you please fix the message in the tool tip to "Remove and stop tracking the DECOMMISSIONED, IN_MAINTENANCE, and DEAD nodes."
> 
> image

@ArafatKhan2198 
Completed changes in latest commit.

**Review by ArafatKhan2198 — APPROVED** [review]
Thanks for the changes @smitajoshi12 
LGTM +1

**Review by devabhishekpal — APPROVED** [review]
Changes look good to me
Thanks @smitajoshi12

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx:466]: This isn't doing what the comments says. Based on the comment you'd like to enable the datanode deletion if either its state is dead or its operational state is decommissioned or in maintenance. 
The code checks if it's dead and it only later checks based on the operational state. E.g. if a datanode is not dead, but in maintenance it won't let it to be removed, but based on the comment and the PR description it should act differently. 
I'm not 100% sure what states can a decommissioned or in maintenance operational stated datanode can be in, but I'd go with a steady solution and check for all three conditions on the same level.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx:525]:
```suggestion
                title={`Are you sure you want Recon to stop tracking the selected ${selectedRowKeys.length} datanodes?`}
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx:531]:
```suggestion
                <Tooltip placement="topLeft" title="Remove the dead or decommissioned or in maintenance datanodes.">
```

**Review by dombizita — CHANGES_REQUESTED** [review]
Thanks for working on this @smitajoshi12! The main part of this change is when you would like users to be able to remove datanodes from the Recon UI. I think that check is not correct currently, please check my comments!

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: @dombizita  Done Changes in latest commit.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: @dombizita 
Done changes in latest commit.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: @dombizita 

Corrected Comments in latest commits.

**Comment by smitajoshi12** [comment]
@dombizita @devmadhuu 

Can we go ahead to merge this PR?

**Review by devmadhuu — APPROVED** [review]
Thanks @smitajoshi12 for improving the patch. LGTM +1

**Comment by smitajoshi12** [comment]
> @smitajoshi12 If DN operational state is `DECOMMISSIONED` or Node Status is `DEAD`, then only we allow, We are also **`NOT`** allowing IN_MAINTENANCE node to be removed. And if a node operational state is `DECOMMISSIONED`, it is not necessarily to be in `DEAD` status and node status may still be `HEALTHY`, so first priority is to check DN operational state should be equal to `DECOMMISSIONED`, then in `OR` second condition should be node status as `DEAD`.
> 
> cc: @krishnaasawa1

@devmadhuu 
Disabled removal for Operation State = IN_MAINTENANCE updated JIRA description and PR details also.

**State= DEAD    Operation State=[Any State] ==> Allow for explicit removal
State= [Any]     Operation State=[Decommissioned] ==> Allow for explicit removal**

Noticed one issue in datanode api State= [Any]     Operation State=[Decommissioned]  removal is happening but some interval of 1 min or 2 mins I can see DECOMMISSIONED Records again.

image

**Comment by devmadhuu** [comment]
> > @smitajoshi12 If DN operational state is `DECOMMISSIONED` or Node Status is `DEAD`, then only we allow, We are also **`NOT`** allowing IN_MAINTENANCE node to be removed. And if a node operational state is `DECOMMISSIONED`, it is not necessarily to be in `DEAD` status and node status may still be `HEALTHY`, so first priority is to check DN operational state should be equal to `DECOMMISSIONED`, then in `OR` second condition should be node status as `DEAD`.
> > cc: @krishnaasawa1
> 
> @devmadhuu Disabled removal for Operation State = IN_MAINTENANCE updated JIRA description and PR details also.
> 
> **State= DEAD Operation State=[Any State] ==> Allow for explicit removal State= [Any] Operation State=[Decommissioned] ==> Allow for explicit removal**
> 
> Noticed one issue in datanode api State= [Any] Operation State=[Decommissioned] removal is happening but some interval of 1 min or 2 mins I can see DECOMMISSIONED Records again.
> 
> image

Created [HDDS-11032](https://issues.apache.org/jira/browse/HDDS-11032) to check this and fix it.

**Comment by smitajoshi12** [comment]
> > > @smitajoshi12 If DN operational state is `DECOMMISSIONED` or Node Status is `DEAD`, then only we allow, We are also **`NOT`** allowing IN_MAINTENANCE node to be removed. And if a node operational state is `DECOMMISSIONED`, it is not necessarily to be in `DEAD` status and node status may still be `HEALTHY`, so first priority is to check DN operational state should be equal to `DECOMMISSIONED`, then in `OR` second condition should be node status as `DEAD`.
> > > cc: @krishnaasawa1
> > 
> > 
> > @devmadhuu Disabled removal for Operation State = IN_MAINTENANCE updated JIRA description and PR details also.
> > **State= DEAD Operation State=[Any State] ==> Allow for explicit removal State= [Any] Operation State=[Decommissioned] ==> Allow for explicit removal**
> > Noticed one issue in datanode api State= [Any] Operation State=[Decommissioned] removal is happening but some interval of 1 min or 2 mins I can see DECOMMISSIONED Records again.
> > image
> 
> Created [HDDS-11032](https://issues.apache.org/jira/browse/HDDS-11032) to check this and fix it.

@devmadhuu 
Updated JIRA and PR Description and Tooltip and comments in code did all testing it is working as expected.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: Correct this comment in code.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: Shouldn't this if check also have "`IN_MAINTENANCE`"  and "`DECOMMISSIONED`" ?

**Review by devmadhuu — COMMENTED** [review]
@smitajoshi12 few more comments.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: In latest commit it was corrected

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: in latest commit it was corrected

**Review by devmadhuu — APPROVED** [review]
Thanks @smitajoshi12  for working on this patch. Changes LGTM +1

**Comment by devmadhuu** [comment]
Thanks @smitajoshi12 for the patch and @dombizita @ArafatKhan2198 @devabhishekpal for review.

**Comment by dombizita** [comment]
@devmadhuu I didn't dismiss my previous [review request](https://github.com/apache/ozone/pull/6411#pullrequestreview-2107317024) yet, please next time wait for that before merging (I'm not even sure how to merge a PR until there is a pending review request)

**Comment by devmadhuu** [comment]
> @devmadhuu I didn't dismiss my previous [review request](https://github.com/apache/ozone/pull/6411#pullrequestreview-2107317024) yet, please next time wait for that before merging (I'm not even sure how to merge a PR until there is a pending review request). As I see there were many changes in the patch since then.

@dombizita I reviewed your comment where you asked to make sure that code and code comments matches, so after @smitajoshi12 updated her comments in code and we changed the condition to check only for DEAD nodes, it was resolved. And since your comments was mainly for code comments and was resolved, we go ahead after all reviews. If you still feel something not resolved as per your comment, kindly raise another JIRA to handle.
cc: @krishnaasawa1

**Comment by dombizita** [comment]
@smitajoshi12 I think it's better if the person who gave the review comment resolves the review comment, as they can decide if the conversation is resolved or not. Please next time don't resolve my comments, even if you did the suggested changes, wait for me to check. Thanks!

**Pull request merged**
